marketplace-service/src/main/java/com/axonivy/market/service/impl/MetadataServiceImpl.java

```diff
@@ -0,0 +1,331 @@
+    this.metadataSyncRepo = metadataSyncRepo;
+    this.productJsonRepo = productJsonRepo;
+    this.mavenArtifactVersionRepo = mavenArtifactVersionRepo;
+    this.metadataRepo = metadataRepo;
+    this.imageService = imageService;
+    this.productModuleContentRepo = productModuleContentRepo;
+    this.productJsonContentService = productJsonContentService;
+  }
+
+  private static void updateMavenArtifactVersionCacheWithModel(MavenArtifactVersion artifactVersionCache,
+      String version, Metadata metadata) {
+    if (metadata.isProductArtifact()) {
+      if (artifactVersionCache.getProductArtifactsByVersion().computeIfAbsent(version,
+          k -> new ArrayList<>()).stream().anyMatch(artifact -> metadata.getName().equals(artifact.getName()))) {
+        return;
+      }
+      artifactVersionCache.getProductArtifactsByVersion().get(version).add(
+          MavenUtils.buildMavenArtifactModelFromSnapShotMetadata(version, metadata));
+    } else {
+      artifactVersionCache.getAdditionalArtifactsByVersion().computeIfAbsent(version, k -> new ArrayList<>()).add(
+          MavenUtils.buildMavenArtifactModelFromSnapShotMetadata(version, metadata));
+    }
+  }
+
+  private void updateMavenArtifactVersionData(Product product, List<String> releasedVersions,
+      Set<Metadata> metadataSet, MavenArtifactVersion artifactVersionCache) {
+    for (Metadata metadata : metadataSet) {
+      String metadataContent = MavenUtils.getMetadataContentFromUrl(metadata.getUrl());
+      if (StringUtils.isBlank(metadataContent)) {
+        continue;
+      }
+      MetadataReaderUtils.parseMetadataFromString(metadataContent, metadata);
+      updateMavenArtifactVersionFromMetadata(artifactVersionCache, metadata);
+      updateContentsFromNonMatchVersions(product.getId(), releasedVersions, metadata);
+    }
+  }
+
+  public void syncAllProductMavenMetadata() {
+    List<Product> products = productRepo.getAllProductWithIdAndReleaseTagAndArtifact();
+    log.warn("**MetadataService: Start to sync version for {} product(s)", products.size());
+    for (Product product : products) {
+      // Set up cache before sync
+      String productId = product.getId();
+      Set<Metadata> metadataSet = new HashSet<>(metadataRepo.findByProductId(product.getId()));
+      MavenArtifactVersion artifactVersionCache = mavenArtifactVersionRepo.findById(product.getId()).orElse(
+          new MavenArtifactVersion(productId, new HashMap<>(), new HashMap<>()));
+      MetadataSync syncCache = metadataSyncRepo.findById(product.getId()).orElse(
+          MetadataSync.builder().productId(product.getId()).syncedTags(new HashSet<>()).build());
+      Set<Artifact> artifactsFromNewTags = new HashSet<>();
+
+      // Find artifacts form unhandled tags
+      List<String> nonSyncedVersionOfTags = VersionUtils.getNonSyncedVersionOfTagsFromMetadataSync(
+          product.getReleasedVersions(),
+          syncCache);
+      if (!CollectionUtils.isEmpty(nonSyncedVersionOfTags)) {
```

Comment:
would be nice if we can use ObjectUtils.isNotEmpty(nonSyncedVersionOfTags) instead of ! of colelctionutil,